### PR TITLE
Add delayed update

### LINKED ISSUE #239: Port delayed update algorithms
1. both CPU and CUDA implementation will be ported.
2. unit tests are ported.

### PR DESCRIPTION
**Pull request by ye-luo**
Closes #239.
Unit test and "check_wfc -f Det" are added.

### CODE REVIEW

**Comment by ye-luo** [comment]
test this please

**Comment by ye-luo** [comment]
test this please

**Comment by ye-luo** [comment]
Test this please

**Comment by ye-luo** [comment]
Test this please

File: src/QMCWaveFunctions/DiracDeterminantRef.h [hunk 1]
```diff
@@ -0,0 +1,150 @@
+//////////////////////////////////////////////////////////////////////////////////////
```
Comment (PDoakORNL) [inline src/QMCWaveFunctions/DiracDeterminantRef.h:1]: I would expect this to be a dead simple reference class, not another copy of your current QMCPACK determinant.

File: src/QMCWaveFunctions/DiracMatrix.h [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+#include "Numerics/OhmmsBlas.h"
+#include "Numerics/OhmmsPETE/OhmmsMatrix.h"
+#include "Numerics/BlasThreadingEnv.h"
+#include "Utilities/scalar_traits.h"
+#include "Utilities/SIMD/algorithm.hpp"
+#include "QMCWaveFunctions/DeterminantHelper.h"
+
+namespace qmcplusplus
+{
+inline void Xgetrf(int n, int m, float* restrict a, int lda, int* restrict piv)
```
Comment (PDoakORNL) [inline src/QMCWaveFunctions/DiracMatrix.h]: One of the things I had to deal with in one code was the practice of just throwing blas wrappers and slightly more elaborate linear algebra functions into another classes header file.  These then just end up lurking in the qmcplusplus namespace of anything that includes this file. Please put them in another header (and preferably a .cpp) contained in a namespace or functor class .

File: src/QMCWaveFunctions/DiracMatrix.h [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+                   std::complex<double>* restrict work,
+                   int& lwork)
+{
+  int status;
+  zgetri(n, a, lda, piv, work, lwork, status);
+}
+
+
+template<typename TIN, typename TOUT>
+inline void TansposeSquare(const TIN* restrict in, TOUT* restrict out, size_t n, size_t lda)
```
Comment (PDoakORNL) [inline src/QMCWaveFunctions/DiracMatrix.h]: spelling

File: src/QMCWaveFunctions/einspline_spo.hpp [hunk 4]
```diff
@@ -155,35 +159,64 @@ struct einspline_spo : public SPOSet
   }
 
   /** evaluate psi */
-  inline void evaluate_v(const PosType& p)
+  inline void evaluate_v(const ParticleSet& P, int iat)
```
Comment (PDoakORNL) [inline src/QMCWaveFunctions/einspline_spo.hpp:162]: I saw it as a big plus that einspline_spo had no knowledge of ParticleSet in Miniqmc.  Why bring it back?

File: src/Utilities/scalar_traits.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+template<typename T>
+struct scalar_traits<std::complex<T>>
+{
+  enum
+  {
+    DIM = 2
+  };
+  typedef T real_type;
+  typedef std::complex<T> value_type;
+  static inline T* get_address(std::complex<T>* a) { return reinterpret_cast<T*>(a); }
```
Comment (PDoakORNL) [inline src/Utilities/scalar_traits.h]: Unless we need this get_address for some c call compatibility I don't think this dark part of the standard need be made easy. Also I find it strange for an "address" to be anything other than `void*` this is more like get_ptr.

**Review by PDoakORNL — CHANGES_REQUESTED** [review]
Good improvements to the build and check_det is gone also good. But import of ParticleSet, EinsplineSPO and DiracDeterminant entanglements not previously an issue in miniqmc. Whether those should be added to develop is a question of what miniqmc is supposed to be...

**Comment by ye-luo** [comment]
I think the added entanglements are necessary for real QMC algorithm and QMCPACK needs.

Ownership of SPO.
When the trial wavefunction is Slater-Jastrow type, it may contains determinants using SPOs. When the whole wavefunction doesn't contain Slater determinant, then there is no SPO. Then SPO needed to be owned by determinants. The code is more modular and composable.
If SPO is owned outside determinants, we have to deal with a null pointer or size 0 vector to handle SPO. This is not a clean route and indicates a non-generic design.

EinsplineSPO or base class SPOSet.
Even though that the miniQMC only supports spline SPO, the design needs to be flexible to be incorporated back to QMCPACK. So hard coding EinsplineSPO and fulfilling only EinsplineSPO needs should be avoided. I'm not saying SPOSet is perfect and believe some call interfaces needs to be updated.

SPOSet using ParticleSet instead of particle positions.
SPOSet represents all the SPOs including LCAO and hybrid representation. Those calculation needs electron ion-distances. ParticleSet owns the particle position and distance tables. Whether to use or not all the data owned by ParticleSet depends on the derived SPO class. I would argue that ParticleSet is better than position only because ParticleSet is more generic.
`SPOSet::evaluate(ParticleSet elecs, int iel)` is more generic than diverged `EinsplineSPO::evaluate(Pos_t R) and
LCAO::evaluate(Pos_t R, Pos_v ions)` or `LCAO::evaluate(Pos_t R, Pos_v displ_ions)`
Passing ParticleSet should not stop you doing anything achievable with positions only.

File: src/QMCWaveFunctions/DiracDeterminantRef.h (hunk 1 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/DiracDeterminantRef.h:1]: This is copied from QMCPACK but trimmed down and updated for miniQMC needs. I believe this is a simple reference class.

File: src/QMCWaveFunctions/DiracMatrix.h (hunk 2 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/DiracMatrix.h]: Yes. This should be moved out. Will fix.

File: src/QMCWaveFunctions/DiracMatrix.h (hunk 3 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/DiracMatrix.h]: Probably we can just delete this function. Will check.

File: src/Utilities/scalar_traits.h (hunk 5 above)
Comment (ye-luo) [reply src/Utilities/scalar_traits.h]: I will have a close look and clean it up.

**Comment by PDoakORNL** [comment]
I don't understand the argument about the harmonic oscillator basis, but if you are just going to appeal to the main source code everytime miniqmc simply isn't useful for design discussions.

I think it's much clearer if the actual terms needed to evaluate the function are in the function signature. If that requires specialized code that is fine there is an important detail that at least the einspline SPO doesn't have.

So to me the second is much better. The idea that `elecs` would somehow change the ion positions used to evaluate the LCAO is counter intuitive to me.

**Comment by PDoakORNL** [comment]
Anyway make the other changes and I will approve, the ParticleSet discussion is too big for this PR.

**Comment by ye-luo** [comment]
I made a mistake harmonic oscillator is a SPO not a WaveFunctionComponent. I correct all the words and removed harmonic oscillator as example.

I would consider the old design as one of the candidates of the new design not being excluded.

I think for virtual functions and functions of templates classes, it is better to have call APIs as generic as possible. If call APIs and classes both need specialization, we cannot solve an all-to-all problem.

File: src/QMCWaveFunctions/DiracDeterminantRef.h (hunk 1 above)
Comment (PDoakORNL) [reply src/QMCWaveFunctions/DiracDeterminantRef.h:1]: To be honest I don't really get the need for the reference class or at least one that is no more straight forward than the actual implementation. A good unit test for DiracDeterminant would be much better.

**Comment by PDoakORNL** [comment]
Seems more like a specialization on an `isDependentOnIonPositions` trait.

File: src/QMCWaveFunctions/DiracDeterminantRef.h (hunk 1 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/DiracDeterminantRef.h:1]: I feel the same. The reference class was a burden for me when I maintain miniQMC.
The up side of keeping a reference is that it demonstrates how two implementation can coexist and checked against each other.
I do feel unit test is way more important and I did port the unit test DiracDeterminant in this PR.

**Comment by ye-luo** [comment]
@PDoakORNL Could you correct my following broken code using `isDependentOnIonPositions` trait?
The following code doesn't compile because EinsplineSPO and LCAO only have one of the two evaluate interface.
Please modify and make it working.
```
Deteriminant<class SPOType>
{
void ratioGrad()
{
   SPOType spo;
   if(isDependentOnIonPositions<SPOType>::value)
      spo.evaluate(Pos_t R, Pos_v ions); // LCAO only has this interface
   else
      spo.evaluate(Pos_t R); // EinsplineSPO only has this interface
}
}
```

**Comment by PDoakORNL** [comment]
```C++
#include<vector>
#include<type_traits>
#include<iostream>

class LCAOSet
{
 public:
    void evaluate(double pos, std::vector<double>& ions) { std::cout << "ions matter\n"; }
};

class EinsplineSet
{
 public:
    void evaluate(double pos) { std::cout << "what's an ion?\n"; }
};

template <typename t>
struct isDependentOnIonPositions : public std::false_type {};

template<>
struct isDependentOnIonPositions<LCAOSet> : public std::true_type {};

template<class SPOT>
struct StateIShouldNotHave;

template<>
struct StateIShouldNotHave<EinsplineSet>
{
    double pos = 1.0;
};

template<>
struct StateIShouldNotHave<LCAOSet>
{
  double pos = 1.0;
  std::vector<double> ions = {1.0, 2.0, 3.0};
};

template<class SPOT>
class Determinant
{
  SPOT spo;
  StateIShouldNotHave<SPOT> state;
 public:
    template<typename U = SPOT,
       typename std::enable_if<isDependentOnIonPositions<U>::value, std::nullptr_t>::type = nullptr>
    void ratioGrad()
    {
      spo.evaluate(state.pos , state.ions);
    }

    template<typename U = SPOT,
       typename std::enable_if<!isDependentOnIonPositions<U>::value, std::nullptr_t>::type = nullptr>
    void ratioGrad()

    {
      spo.evaluate(state.pos);
    }
};


int main()
{
  Determinant<LCAOSet> det_lcao;
  Determinant<EinsplineSet> det_einsplineset;

  det_einsplineset.ratioGrad();
  det_lcao.ratioGrad();
}
```

Even xlc++ can do it

Edit: I'd rather the enable_if in the parameter list

File: src/QMCWaveFunctions/DiracMatrix.h (hunk 2 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/DiracMatrix.h]: All moved to OhmmsBlas.h and did a bit cleaning.

File: src/QMCWaveFunctions/DiracMatrix.h (hunk 3 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/DiracMatrix.h]: Deleted.

File: src/Utilities/scalar_traits.h (hunk 5 above)
Comment (ye-luo) [reply src/Utilities/scalar_traits.h]: Just deleted them.

**Review by PDoakORNL — APPROVED** [review]

**Comment by ye-luo** [comment]
@PDoakORNL Thank you. I expected this complexity. If I have not just one call inside ratioGrad but very long function body, having it repeated the whole function because of 1 call difference, I feel, is not really worth it. If we have a third SPOSet which need another interface, the ratioGrad need a third version. Hopefully you can consider evaluate(const ParticleSet&) and we don't need to maintain this complexity.

The code pattern you showed is very powerful to handle data types or functions from libraries which is often not allowed to be modified. Having traits and SFINAE works very well.

**Pull request merged**